test/widgets/test_windowtabs.py

```diff
@@ -0,0 +1,162 @@
+from libqtile import bar, layout, widget
+from libqtile.config import Screen
+from libqtile.confreader import Config
+
+
+class FakeScreenConfig(Config):
+    auto_fullscreen = True
+    groups = [
+        libqtile.config.Group("a"),
+        libqtile.config.Group("b")
+    ]
+    layouts = [
+        layout.Stack()
+    ]
+    floating_layout = libqtile.resources.default_config.floating_layout
+    keys = []
+    mouse = []
+    fake_screens = [
+        Screen(
+            top=bar.Bar(
+                [
+                    widget.WindowTabs(),
+                ],
+                24,
+            ),
+            bottom=bar.Bar(
+                [
+                    widget.WindowTabs(selected="!!"),
+                ],
+                24,
+            ),
+            x=0, y=0, width=900, height=960,
+        ),
+    ]
+    screens = []
+
+
+xephyr_config = {
+    "xinerama": False,
+    "two_screens": False,
+    "width": 900,
+    "height": 960
+}
+fakescreen_config = pytest.mark.parametrize("xephyr, manager", [(xephyr_config, FakeScreenConfig)], indirect=True)
+
+
+@fakescreen_config
+def test_single_window_states(manager):
+
+    def widget_text():
+        return manager.c.bar["top"].info()["widgets"][0]["text"]
+
+    # Default _TextBox text is " " and no hooks fired yet.
+    assert widget_text() == " "
+
+    # Load xeyes
+    proc = manager.test_xeyes()
+    assert widget_text() == "<b>xeyes</b>"
+
+    # Maximize window
+    manager.c.window.toggle_maximize()
+    assert widget_text() == "<b>[] xeyes</b>"
+
+    # Minimize window
+    manager.c.window.toggle_minimize()
+    assert widget_text() == "<b>_ xeyes</b>"
+
+    # Float window
+    manager.c.window.toggle_minimize()
+    manager.c.window.toggle_floating()
+    assert widget_text() == "<b>V xeyes</b>"
+
+    # Kill xeyes and check text again
+    # NB hooks fired so empty string is now ""
+    manager.kill_window(proc)
+    assert widget_text() == ""
+
+
+@fakescreen_config
+def test_multiple_windows(manager):
+
+    def widget_text():
+        return manager.c.bar["top"].info()["widgets"][0]["text"]
+
+    window_one = manager.test_window("one")
+    assert widget_text() == "<b>one</b>"
+
+    window_two = manager.test_window("two")
+    assert widget_text() in ["<b>two</b> | one", "one | <b>two</b>"]
+
+    manager.c.layout.next()
+    assert widget_text() in ["<b>one</b> | two", "two | <b>one</b>"]
+
+    manager.kill_window(window_one)
+    assert widget_text() == "<b>two</b>"
+
+    manager.kill_window(window_two)
+    assert widget_text() == ""
+
+
+@fakescreen_config
+def test_selected(manager):
+
+    # Bottom bar widget has custom "selected" indicator
+    def widget_text():
+        return manager.c.bar["bottom"].info()["widgets"][0]["text"]
+
+    window_one = manager.test_window("one")
+    assert widget_text() == "!!one!!"
+
+    manager.kill_window(window_one)
+    assert widget_text() == ""
+
+
+# This test is needed to hit 100% coverage
```

Comment:
Feels like you're right and we can just drop the "None" case, a window in the windows list should never be none.